xl-v1-utils.go

```diff
@@ -24,6 +24,72 @@ import (
 	"time"
 )
 
+// error based on total errors and quorumCount.
+func reduceError(errs []error, quorumCount int) error {
+	var (
+		diskNotFoundCount       int
+		volumeNotFoundCount     int
+		volumeAccessDeniedCount int
+		fileNotFoundCount       int
+		fileAccessDeniedCount   int
+		fileLongNameCount       int
+	)
+	// Count different types of concocted errors.
+	for _, err := range errs {
+		if err == errFileNotFound {
+			fileNotFoundCount++
+		} else if err == errFileAccessDenied {
+			fileAccessDeniedCount++
+		} else if err == errFileNameTooLong {
+			fileLongNameCount++
+		} else if err == errDiskNotFound {
+			diskNotFoundCount++
+		} else if err == errVolumeAccessDenied {
+			volumeAccessDeniedCount++
+		} else if err == errVolumeNotFound {
+			volumeNotFoundCount++
+		}
+	}
+
+	switch {
+	case diskNotFoundCount == len(errs):
+		// If we have errors with disk not found equal to the
+		// number of disks, return as errDiskNotFound.
+		return errDiskNotFound
+	case volumeAccessDeniedCount == len(errs):
+		// If we have errors with disks access denied count equal to the
+		// number of errors, return as errVolumeAccessDenied.
+		return errVolumeAccessDenied
+	}
+	// If we have errors with 'file not found' greater than
+	// quorumCount, return as errFileNotFound.
+	// else if we have errors with 'volume not found'
+	// greater than quorumCount, return as errVolumeNotFound.
+	if fileNotFoundCount+diskNotFoundCount > len(errs)-quorumCount {
+		return errFileNotFound
+	} else if fileLongNameCount+diskNotFoundCount > len(errs)-quorumCount {
+		return errFileNameTooLong
+	} else if volumeNotFoundCount+diskNotFoundCount > len(errs)-quorumCount {
+		return errVolumeNotFound
+	} else if fileAccessDeniedCount+diskNotFoundCount > len(errs)-quorumCount {
+		return errFileAccessDenied
+	}
+	return nil
```

Comment:
still does not cover all cases, for example in 4x4 if fileNotFoundCount is 3 and volumeNotFoundCount is 2, reduceError() should return error but it returns nil